akd_core/src/verify/history.rs

```diff
@@ -193,6 +222,208 @@ pub fn key_history_verify<TC: Configuration>(
     Ok(results)
 }
 
+fn verify_with_history_params(
+    current_epoch: u64,
+    akd_label: &AkdLabel,
+    proof: &HistoryProofV2,
+    params: HistoryParams,
+) -> Result<(Vec<u64>, Vec<u64>), VerificationError> {
+    let num_proofs = proof.update_proofs.len();
+
+    // Make sure the update proofs are non-empty
+    if num_proofs == 0 {
+        return Err(VerificationError::HistoryProof(format!(
+            "No update proofs included in the proof of user {akd_label:?} at epoch {current_epoch:?}!"
+        )));
+    }
+
+    // Check that the sent proofs are for a contiguous sequence of decreasing versions
+    for count in 0..num_proofs {
+        if count > 0 {
+            // Make sure this proof is for a version 1 more than the previous one.
+            if proof.update_proofs[count].version + 1 != proof.update_proofs[count - 1].version {
+                return Err(VerificationError::HistoryProof(format!(
+                    "Update proofs should be ordered consecutively and in decreasing order.
+                    Error detected with version {} = {}, followed by version {} = {}",
+                    count,
+                    proof.update_proofs[count].version,
+                    count - 1,
+                    proof.update_proofs[count - 1].version
+                )));
+            }
+        }
+    }
+
+    let mut start_version = proof.update_proofs[0].version;
+    let mut end_version = proof.update_proofs[0].version;
+    proof.update_proofs.iter().for_each(|update_proof| {
+        if update_proof.version < start_version {
+            start_version = update_proof.version;
+        }
+        if update_proof.version > end_version {
+            end_version = update_proof.version;
+        }
+    });
+
+    if start_version == 0 {
+        return Err(VerificationError::HistoryProof(
+            "Computed start version for the key history should be non-zero".to_string(),
+        ));
+    }
+
+    if end_version > current_epoch {
+        return Err(VerificationError::HistoryProof(
+            "Computed end version for the key history should not exceed current epoch".to_string(),
+        ));
+    }
+
+    match params {
+        HistoryParams::Complete => {
+            // Make sure the start version is 1
+            if start_version != 1 {
+                return Err(VerificationError::HistoryProof(format!(
+                    "Expected start version to be 1 given that it is a complete history, but got start_version = {}",
+                    start_version
+                )));
+            }
+        }
+        HistoryParams::MostRecent(recency) =>
+        {
+            #[allow(clippy::comparison_chain)]
```

Comment:
Since this match block mostly seems to be aimed at surfacing errors which violate invariants, I think you can add all of these if-checks into arms in the match with corresponding guards to return an `Err`. That may help with the linter warning here, as I think you'd be forced to handle the "ok" case.